### PR TITLE
SCP-1826 Verified renaming and substitution for NEAT types via agda2hs

### PR DESCRIPTION
**Pull request by jmchapman**
This PR yanks out the definition of well-scoped types, type renaming and substitution used in the NEAT generator into an Agda file and then proves the functor and monad laws for renaming and substitution. We can then compile the Agda file back to Haskell using `agda2hs`.

The Haskell code remains unchanged except `Type.hs` is split into `Term` (not verified) and `Type` (verified). I wrote my own instance of renaming rather than using `deriving Functor` for `TypeG` and I had to kill the import list. I also turned `Common.hs` into an Agda file to get my hands on the datatypes `S` (aka `Maybe`) and `Z` (aka `Empty`).

The NEAT generator is supposed to generate only well-typed programs. It is a model of PLC. Its typechecker makes use of type substitution, which makes use of renaming.. These properties are easy to prove but hard to test and currently untested.

I have committed both the Agda code and the Haskell code.

Some questions:
1. Do we want to commit both Agda and generated Haskell code?
2. Do we want this at all?

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested
- If you updated any cabal files or added Haskell packages:
    - [x] `nix-shell shell.nix --run updateMaterialized` to update the materialized Nix files
    - [ ] Update `hie-*.yaml` files if needed
- If you changed any Haskell files:
    - [x] `nix-shell shell.nix --run fix-stylish-haskell` to fix any formatting issues
- If you changed any Purescript files:
    - [x] `nix-shell shell.nix --run fix-purty` to fix any formatting issues

Pre-merge checklist:
- [ ] Someone approved it
- [x] Commits have useful messages
- [ ] Review clarifications made it into the code
- [x] History is moderately tidy; or going to squash-merge

### CODE REVIEW

**Comment by michaelpj** [comment]
I think this is cool, and it's a good testbed for actually using something generated by `agda2hs` - we'll need to do something like this if we wanted to have e.g. verified evaluators using it.

WRT the generation, I think we have two options:
- Serious version: require `agda2hs` as a build tool for `plutus-core`, add a custom setup to generate the files. Or figure out some other way of doing the code generation.
    - It would be nice to have a better code generation method but I'm not sure whether there are any...
- Easy version: just commit the generated code for now.
    - The big downsides are that we won't be checking the Agda code and regenerating things will be annoyingly manual.

I *think* we should probably stick with the easy version for now. Questions:
- Is `agda2hs` on Hackage?
- Can we get `agda2hs` to generate a big scary comment at the top of generated modules saying "This is a generated module, do not edit manually!" or something?

**Comment by jmchapman** [comment]
> I think this is cool, and it's a good testbed for actually using something generated by `agda2hs` - we'll need to do something like this if we wanted to have e.g. verified evaluators using it.

Great!

> I think we should probably stick with the easy version for now. Questions:

> * Is agda2hs on Hackage?

No. Perhaps we should consider a release of it. And then it might make sense to do the _serious_ version.

> * Can we get agda2hs to generate a big scary comment at the top of generated modules saying "This is a generated module, do not edit manually!" or something?

I wrote a warning into each file as a Haskell comment and wrote an Agda comment at the top.

**Review by effectfully — APPROVED** [review]
That looks really cool. I wonder if it makes incorporating extensible builtins a harder task, though. I guess we can't embed stuff from [`Universe.Core`](https://github.com/input-output-hk/plutus/blob/29d4f0d2297b9c19422077bbce8b1955e81ce4bc/plutus-core/src/Language/PlutusCore/Universe/Core.hs) in a transparent manner, so that you can for example match on Haskell's `Some` in Agda? I.e. will we need to (partially) reimplement definitions from `Universe.Core` in Agda? Not a big deal, I guess.

**Comment by jmchapman** [comment]
> That looks really cool. I wonder if it makes incorporating extensible builtins a harder task, though. I guess we can't embed stuff from Universe.Core in a transparent manner, so that you can for example match on Haskell's Some in Agda? I.e. will we need to (partially) reimplement definitions from Universe.Core in Agda? Not a big deal, I guess.

I think you're right that one would have to replicate `Universe.Core` as you cannot currently compile it with agda2hs.

I would like to go ahead with this which is quite self contained and we can back it out later if necessary.

**Comment by jmchapman** [comment]
@michaelpj do you have any further thoughts?

File: plutus-core/generators/Language/PlutusCore/Generators/NEAT/Type.agda [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+-- warning to be written to Haskell file:
+{-# FOREIGN AGDA2HS
+{-
+!!! THIS FILE IS GENERATED FROM Type.agda
+!!! DO NOT EDIT THIS FILE. EDIT Type.agda
+!!! AND THEN RUN agda2hs ON IT.
+-}
+#-}
+
+{-# FOREIGN AGDA2HS
```
Comment (michaelpj) [inline plutus-core/generators/Language/PlutusCore/Generators/NEAT/Type.agda:16]: How does this parse??

**Comment by michaelpj** [comment]
Seems fine, merge at your leisure!

File: plutus-core/generators/Language/PlutusCore/Generators/NEAT/Type.agda (hunk 1 above)
Comment (jmchapman) [reply plutus-core/generators/Language/PlutusCore/Generators/NEAT/Type.agda:16]: oops, missed that. I would have to defer to Ulf for that

**Pull request merged**